### PR TITLE
[ALSP] Implements hearbeat and decay logic for ALSP manager

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This Pull Request (PR) signifies the fourth installment of the Reputation Management System (RMS) within the scope of the Application Layer Spam Prevention (ALSP) framework, specifically designed for the Flow blockchain. Expanding upon the work carried out in the preceding PRs, it incorporates a consistent heartbeat mechanism within the ALSP `manager`.

This mechanism systematically scans through the spam records and mitigates the penalty points of the nodes. It performs this operation by transitioning from a negative score back to neutral zero via progressive decay steps. The current default rate for this decaying process is set at `1000`. This implies that the penalty score of a spamming peer is consistently reduced back to zero through an increment of `+1000` every second. When the penalty score reduces to `zero`, it remains unchanged.

Future PRs aim to introduce a customized decay rate for each node, which would be calibrated according to the node's historical spamming behavior.

**Note**: To optimize the review and integration processes, this PR has been crafted to be concise and self-sufficient. As the development of RMS constitutes a substantial endeavor, we have opted to break down its implementation into a series of smaller, independent PRs. This strategy not only accelerates the review and development processes but also allows us to obtain valuable feedback early in the development stages.

### CODE REVIEW

File: cmd/node_builder.go [hunk 1]
```diff
@@ -220,6 +220,10 @@ type AlspConfig struct {
 	// This is useful for managing production incidents.
 	// Note: under normal circumstances, the ALSP module should not be disabled.
 	DisablePenalty bool
+
+	// HearBeatInterval is the interval between heartbeats sent by the ALSP module. The heartbeats are recurring
```
Comment (peterargue) [inline cmd/node_builder.go:224]:
```suggestion
	// HeartBeatInterval is the interval between heartbeats sent by the ALSP module. The heartbeats are recurring
```

File: network/alsp/manager/manager.go [hunk 2]
```diff
@@ -171,6 +184,11 @@ func NewMisbehaviorReportManager(cfg *MisbehaviorReportManagerConfig, opts ...Mi
 		store,
 		m.processMisbehaviorReport).Build()
 
+	m.workerPool = worker.NewWorkerPoolBuilder[internal.ReportedMisbehaviorWork](
+		cfg.Logger,
+		store,
+		m.processMisbehaviorReport).Build()
+
```
Comment (peterargue) [inline network/alsp/manager/manager.go:191]: this is duplicated
```suggestion
```

File: network/alsp/manager/manager.go [hunk 3]
```diff
@@ -237,6 +259,77 @@ func (m *MisbehaviorReportManager) HandleMisbehaviorReport(channel channels.Chan
...
+			// is a positive number. So the penalty is getting closer to zero.
+			// We use math.Min() to make sure the penalty is never positive.
+			record.Penalty = math.Min(record.Penalty+record.Decay, 0)
+			return record, nil
+		})
+
+		// any error here is fatal because it indicates a bug in the cache. All ids being iterated over are in the cache,
+		// and adjust function above should not return an error unless there is a bug.
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to decay spam record %x: %w", id, err))
```
Comment (peterargue) [inline network/alsp/manager/manager.go:322]: it's a bit pedantic, but I think `onHeartbeat` should return an error, and the throw should happen in the `startHeartbeatTicks` loop

File: network/alsp/manager/manager.go [hunk 4]
```diff
@@ -237,6 +259,77 @@ func (m *MisbehaviorReportManager) HandleMisbehaviorReport(channel channels.Chan
...
+// are used to perform periodic tasks.
+// Args:
+//
+//	ctx: the context.
+//	interval: the interval between two ticks.
+//
+// Returns:
+//
+//	none.
+func (m *MisbehaviorReportManager) startHeartbeatTicks(ctx irrecoverable.SignalerContext, interval time.Duration) {
```
Comment (peterargue) [inline network/alsp/manager/manager.go:273]: the `start` terminology makes it sound like this is non-blocking. How about calling it a loop?
```suggestion
func (m *MisbehaviorReportManager) heartbeatLoop(ctx irrecoverable.SignalerContext, interval time.Duration) {
```

**Review by peterargue — APPROVED** [review]
added a few small comments, but otherwise looks good.

**Review by kc1116 — APPROVED** [review]

File: network/alsp/manager/manager.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/alsp/manager/manager.go]: https://github.com/onflow/flow-go/pull/4367/commits/b46a2b4a0d0fdab50a3da46208085ae8ead4842b

**Comment by yhassanzadeh13** [comment]
bors merge

**Comment by yhassanzadeh13** [comment]
bors merge

**Comment by yhassanzadeh13** [comment]
bors merge

**Comment by yhassanzadeh13** [comment]
bors merge

**Comment by yhassanzadeh13** [comment]
bors merge

**Pull request merged**